src/Services/Azure/Authentication/CustomChainedCredential.cs

```diff
@@ -119,4 +135,76 @@ private static DefaultAzureCredential CreateDefaultCredential(string? tenantId)
 
         return new DefaultAzureCredential(defaultCredentialOptions);
     }
+
+    private static TokenCredential? CreateVsCodeBrokerCredential(string? tenantId)
+    {
+        const string vsCodeClientId = "aebc6443-996d-45c2-90f0-388ff96faa56";
+        string userProfile = Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);
+        string authRecordPath = Path.Combine(userProfile, ".azure", "ms-azuretools.vscode-azureresourcegroups", "authRecord.json");
+        if (!File.Exists(authRecordPath))
+        {
+            // Try .Azure if .azure is not present
+            authRecordPath = Path.Combine(userProfile, ".Azure", "ms-azuretools.vscode-azureresourcegroups", "authRecord.json");
+            if (!File.Exists(authRecordPath))
+            {
+                return null;
+            }
+        }
+
+        AuthenticationRecord? authRecord;
+        try
+        {
+            using var stream = File.OpenRead(authRecordPath);
+            authRecord = AuthenticationRecord.Deserialize(stream);
+        }
+        catch
+        {
+            // Deserialization failed
+            return null;
+        }
+
+        if (authRecord is null)
+            return null;
+
+        // Validate client ID
+        if (!string.Equals(authRecord.ClientId, vsCodeClientId, StringComparison.OrdinalIgnoreCase))
+            return null;
+
+        // Validate tenant ID if present
+        if (!string.IsNullOrEmpty(authRecord.TenantId) && !IsValidTenantId(authRecord.TenantId))
+            return null;
+
+        // Prefer explicit tenantId, else use from auth record
+        string? effectiveTenantId = !string.IsNullOrEmpty(tenantId)
```

Comment:
>string? effectiveTenantId [](http://example.com/codeflow?start=8&length=25)

Nit: declare this variable as `string` not `string?`. Using `string?` says that the variable could be `null`, but we're doing logic here that ensures the value is not `null`. That means we can use `string` and the compiler will take that into account for finding logical errors.